provider/kubernetes/client.go

```diff
@@ -150,129 +122,84 @@ func (c *clientImpl) WatchAll(namespaces Namespaces, labelSelector string, stopC
 		c.isNamespaceAll = true
 	}
 
-	var informManager informerManager
+	eventHandler := newResourceEventHandler(eventCh)
 	for _, ns := range namespaces {
-		ns := ns
-		informManager.extend(c.WatchIngresses(ns, kubeLabelSelector, eventCh), true)
-		informManager.extend(c.WatchObjects(ns, kindServices, &corev1.Service{}, c.svcStores, eventCh), true)
-		informManager.extend(c.WatchObjects(ns, kindEndpoints, &corev1.Endpoints{}, c.epStores, eventCh), true)
-		// Do not wait for the Secrets store to get synced since we cannot rely on
-		// users having granted RBAC permissions for this object.
-		// https://github.com/containous/traefik/issues/1784 should improve the
-		// situation here in the future.
-		informManager.extend(c.WatchObjects(ns, kindSecrets, &corev1.Secret{}, c.secStores, eventCh), false)
-	}
-
-	var wg sync.WaitGroup
-	for _, informer := range informManager.informers {
-		informer := informer
-		safe.Go(func() {
-			wg.Add(1)
-			informer.Run(stopCh)
-			wg.Done()
+		factory := informers.NewFilteredSharedInformerFactory(c.clientset, resyncPeriod, ns, func(opts *metav1.ListOptions) {
+			opts.LabelSelector = labelSelector
 		})
+		factory.Extensions().V1beta1().Ingresses().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Services().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Endpoints().Informer().AddEventHandler(eventHandler)
+		factory.Core().V1().Secrets().Informer().AddEventHandler(eventHandler)
+		c.factories[ns] = factory
 	}
 
-	if !cache.WaitForCacheSync(stopCh, informManager.syncFuncs...) {
-		return nil, fmt.Errorf("timed out waiting for controller caches to sync")
+	for _, ns := range namespaces {
+		c.factories[ns].Start(stopCh)
 	}
 
-	safe.Go(func() {
-		<-stopCh
-		wg.Wait()
-		close(eventCh)
```

Comment:
@timoreimann  What would happen if we don't close `eventCh` gracefully?

Reply:
From what I can tell, we should be okay not closing it: The latest refactorings seem to have made it obsolete.

I'll do some extra manual tests once the PR is complete.